### PR TITLE
Implement Newton-CG solver

### PR DESCRIPTION
**Pull request by avolkov-intel**
# Description
Add new optimization method newton-cg (or hessian free) as primitive.
This method is an analogue for this optimization method from scipy https://docs.scipy.org/doc/scipy/reference/optimize.minimize-newtoncg.html

Changes proposed in this pull request:
-Implement conjugate gradients solver
-Add tests for cg-solver
-Develop function wrapper interface
-Develop newton-cg optimization primitive
-Develop line search method for 1d-optimization https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.line_search.html
-Add tests for newton-cg solver

### CODE REVIEW

File: cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    const std::int64_t n = x.get_dimension(0);
+    auto* x_ptr = x.get_mutable_data();
+    auto* y_ptr = y.get_mutable_data();
+    sycl::event fill_res_event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.single_task([=]() {
+            *res_gpu = 0;
+        });
+    });
+    fill_res_event.wait_and_throw();
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp:55]: What about passing event to the next kernel?

File: cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+    auto compute_r0_event = element_wise(queue,
+                                         kernel_minus,
+                                         residual,
+                                         b,
+                                         residual,
+                                         { compute_ax0_event }); // r0 = Ax0 - b
+    // compute_r0_event.wait_and_throw();
+    auto tmp_gpu = ndarray<Float, 1>::empty(queue, { 1 }, sycl::usm::alloc::device);
+    auto tmp_ptr = tmp_gpu.get_mutable_data();
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp:115]:
```suggestion
    auto* const  tmp_ptr = tmp_gpu.get_mutable_data();
```

File: cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                                           init_conj_kernel,
+                                           residual,
+                                           conj_vector,
+                                           conj_vector,
+                                           { compute_r0_event }); // p0 = -r0 + 0 * p
+    auto conj_host = conj_vector.to_host(queue, {});
+    dot_product<Float>(queue, residual, residual, tmp_ptr, &r_norm, { compute_r0_event })
+        .wait_and_throw(); // compute r^T r
+
+    for (std::int32_t iter_num = 0; iter_num < maxiter; ++iter_num) {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp:135]: 1) Please use `std::int64_t`
2) Please consider creation of `sycl::event last = ...` and assignment on each iteration

**Comment by samir-nasibli** [comment]
@avolkov-intel please share your sources in comment block's at least used for the current implementation.

File: cpp/oneapi/dal/backend/primitives/optimizers/common.hpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                    Float* res_host,
+                    const event_vector& deps = {});
+
+template <typename Float>
+sycl::event dot_product(sycl::queue& queue,
+                        const ndview<Float, 1>& x,
+                        const ndview<Float, 1>& y,
+                        Float* res_gpu,
+                        Float* res_host,
+                        const event_vector& deps = {});
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/backend/primitives/optimizers/common.hpp:39]: I think these features should be common not only for optimizers

File: cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    auto compute_ax0_event = mul_operator(x, residual, deps); // r = Ax0
+
+    auto compute_r0_event = element_wise(queue,
+                                         kernel_minus,
+                                         residual,
+                                         b,
+                                         residual,
+                                         { compute_ax0_event }); // r0 = Ax0 - b
+    // compute_r0_event.wait_and_throw();
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp:65]:
```suggestion
```

File: cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    queue
+        .submit([&](sycl::handler& cgh) {
+            cgh.depends_on(fill_res_event);
+            const auto range = make_range_1d(n);
+            auto sum_reduction = sycl::reduction(res_gpu, sycl::plus<>());
+            cgh.parallel_for(range, sum_reduction, [=](sycl::id<1> idx, auto& sum) {
+                sum += x_ptr[idx] * y_ptr[idx];
+            });
+        })
+        .wait_and_throw();
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp:56]: What is the reason calling blocks and waits for this tasks to finish?

File: cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        .submit([&](sycl::handler& cgh) {
+            cgh.depends_on(fill_res_event);
+            const auto range = make_range_1d(n);
+            auto sum_reduction = sycl::reduction(res_gpu, sycl::plus<>());
+            cgh.parallel_for(range, sum_reduction, [=](sycl::id<1> idx, auto& sum) {
+                Float val = x_ptr[idx];
+                sum += val >= 0 ? val : -val;
+            });
+        })
+        .wait_and_throw();
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp:87]:
```suggestion
```

File: cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                                           conj_vector,
+                                           conj_vector,
+                                           { compute_r0_event }); // p0 = -r0 + 0 * p
+    auto conj_host = conj_vector.to_host(queue, {});
+    dot_product<Float>(queue, residual, residual, tmp_ptr, &r_norm, { compute_r0_event })
+        .wait_and_throw(); // compute r^T r
+
+    for (std::int64_t iter_num = 0; iter_num < maxiter; ++iter_num) {
+        l1_norm<Float>(queue, residual, tmp_ptr, &r_l1_norm, { compute_conj_event })
+            .wait_and_throw(); // compute norm(residual)
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp:90]: I believe you can do it asynchronously. Please just use an external (relative to the loop) variable

File: cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+        if (r_l1_norm < threshold) {
+            break;
+        }
+        auto compute_matmul_event =
+            mul_operator(conj_vector, buffer, { compute_conj_event }); // compute A p_i
+        dot_product<Float>(queue, conj_vector, buffer, tmp_ptr, &alpha, { compute_matmul_event })
+            .wait_and_throw(); // compute p_i^T A p_i
+        ONEDAL_ASSERT(alpha > 1e-15);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/cg_solver_dpc.cpp:99]: Looks weird.

File: cpp/oneapi/dal/backend/primitives/optimizers/common.hpp [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        ONEDAL_ASSERT(A_.get_dimension(1) == vec.get_dimension(0));
+        ONEDAL_ASSERT(out.get_dimension(0) == vec.get_dimension(0));
+        sycl::event fill_out_event = fill<Float>(q_, out, Float(0), deps);
+        return gemv(q_, A_, vec, out, Float(1), Float(0), { fill_out_event });
+    }
+
+private:
+    sycl::queue q_;
+    const ndview<Float, 2> A_;
+};
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/common.hpp:67]: Please move concrete function implementations into the `.cpp` file

File: cpp/oneapi/dal/backend/primitives/optimizers/common.hpp [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    virtual BaseMatrixOperator<Float>& get_hessian_product() = 0;
+    virtual sycl::event update_x(const ndview<Float, 1>& x,
+                                 bool needHessp = false,
+                                 const event_vector& deps = {}) = 0;
+};
+
+// f(x) = 1/2 x^t A x - b^t x
+// df / dx = Ax - b
+// df / d^2x = A
+template <typename Float>
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/common.hpp:84]: I believe this function is needed only for testing purposes. I insist on moving it into the test compiled files

File: cpp/oneapi/dal/backend/primitives/optimizers/common.hpp [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        return hessp_;
+    }
+
+    sycl::event update_x(const ndview<Float, 1>& x,
+                         bool needHessp = false,
+                         const event_vector& deps = {}) override {
+        auto fill_gradient_event = fill<Float>(q_, gradient_, Float(0), deps);
+        auto fill_value_event = fill<Float>(q_, tmp_, Float(0), deps);
+        auto gemv_event =
+            gemv(q_, A_, x, gradient_, Float(1), Float(0), { fill_gradient_event }); // Ax
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/common.hpp:117]: Please just create `constexpr Float zero(0), one(1);`

File: cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+                        const event_vector& deps) {
+    const std::int64_t n = x.get_dimension(0);
+    auto* const x_ptr = x.get_mutable_data();
+    auto* const y_ptr = y.get_mutable_data();
+    sycl::event fill_res_event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.single_task([=]() {
+            *res_gpu = 0;
+        });
+    });
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp:44]: Use `sycl::fill` instead

File: cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        cgh.fill(res_gpu, Float(0), 1);
+    });
+
+    auto reduction_event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(fill_res_event);
+        const auto range = make_range_1d(n);
+        auto sum_reduction = sycl::reduction(res_gpu, sycl::plus<>());
+        cgh.parallel_for(range, sum_reduction, [=](sycl::id<1> idx, auto& sum) {
+            Float val = x_ptr[idx];
+            sum += val >= 0 ? val : -val;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp:81]:
```suggestion
            sum += 0 < val ? -val : val;
```

File: cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        cgh.depends_on(deps);
+        cgh.fill(res_gpu, Float(0), 1);
+    });
+
+    auto reduction_event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(fill_res_event);
+        const auto range = make_range_1d(n);
+        auto sum_reduction = sycl::reduction(res_gpu, sycl::plus<>());
+        cgh.parallel_for(range, sum_reduction, [=](sycl::id<1> idx, auto& sum) {
+            Float val = x_ptr[idx];
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/common_dpc.cpp:80]:
```suggestion
            const Float val = x_ptr[idx];
```

File: cpp/oneapi/dal/backend/primitives/optimizers/line_search.hpp [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+/// @param deps
+/// @return
+template <typename Float>
+Float backtracking(sycl::queue queue,
+                   BaseFunction<Float>& f,
+                   const ndview<Float, 1>& x,
+                   const ndview<Float, 1>& direction,
+                   ndview<Float, 1>& result,
+                   Float alpha = 1,
+                   const Float c1 = 1e-4,
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/line_search.hpp:47]:
```suggestion
                   Float c1 = 1e-4,
```

File: cpp/oneapi/dal/backend/primitives/optimizers/line_search_dpc.cpp [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        precompute = f.update_x(x, false, deps);
+    }
+    Float f0 = f.get_value();
+    auto grad_f0 = f.get_gradient();
+    Float df0 = 0;
+    dot_product(queue, grad_f0, direction, result.get_mutable_data(), &df0, deps + precompute)
+        .wait_and_throw();
+    bool is_first_iter = true;
+    Float cur_val = 0;
+    while (is_first_iter || cur_val > f0 + c1 * alpha * df0) {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/optimizers/line_search_dpc.cpp:45]: Please control the number of iterations

**Review by KulikovNikita — APPROVED** [review]

**Comment by samir-nasibli** [comment]
I don't understand why we merging PR and ignoring red CI check?

**Comment by samir-nasibli** [comment]
@avolkov-intel @KulikovNikita could you please update?

**Pull request merged**